Support multiple crawler versions

tw4l opened the pull request:
Fixes #1385 

## Changes
Supports multiple crawler 'channels' which can be configured to different browsertrix-crawler versions
- Replaces `crawler_image` in helm chart with `crawler_channels` array similar to how storages are handled
- The `default` crawler channel must always be provided and specifies the default crawler image
- Adds backend `/orgs/{oid}/crawlconfigs/crawler-channels` API endpoint to fetch information about available crawler versions (name, image, and label) and test
- Adds crawler channel select to workflow creation/edit screens and profile creation dialog, and updates related API endpoints and configmaps accordingly. The select dropdown is shown only if more than one channel is configured.
- Adds `crawlerChannel` to workflow and crawl details.
- Add `image` to crawler image, used to display actual image used as part of the crawl.
- Modifies `crawler_crawl_id` backend test fixture to use `test` crawler version to ensure crawler versions other than latest work
- Adds migration to add `crawlerChannel` set to `default` to existing workflow and profile objects and workflow configmaps

## Screenshots

#### Workflow create/edit

<img width="1166" alt="Screen Shot 2023-12-14 at 12 39 56 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/48680e18-ca97-42b8-98a0-75e7c41e76e6">

#### Workflow detail

<img width="1172" alt="Screen Shot 2023-12-14 at 12 40 17 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/ca1a72b9-c127-4fb1-9070-7ee2c3c6b351">

#### Crawl detail

<img width="1178" alt="Screen Shot 2023-12-14 at 12 49 15 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/58a1eaa7-6ea4-4bc8-9550-b0f82f387b4c">

#### Create profile browser dialog

<img width="538" alt="Screen Shot 2023-12-14 at 12 40 27 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/a9e1ac44-86bd-47cb-9b48-7e0aa941c243">

tw4l committed:
Implement multiple crawler versions on backend

Versions are configured in helm chart. Crawls use the crawler
version specified by id in the workflow's crawlerid field. If a
crawlerid is not specified, it defaults to "latest".

tw4l committed:
Add CRAWLER_IMAGE to workflow configmap, fix scheduled crawls

tw4l committed:
Add missing model docstrings

tw4l committed:
Add pylint disable comment

tw4l committed:
Add support for crawler versions in profile browsers

tw4l committed:
Fix typo

tw4l committed:
Add crawler_versions to local yaml example and add comment

tw4l committed:
WIP: Add crawler select to workflow editor

tw4l committed:
Getting closer on workflow frontend

tw4l committed:
Rename component to btrix-select-crawler

tw4l committed:
Rename file

tw4l committed:
Import component

tw4l committed:
Still working on crawler select for workflows

tw4l committed:
Remove console logging

tw4l committed:
Formatting/clean up

tw4l committed:
Add crawler-select to new profile browser form

tw4l committed:
Add crawlerid to profile

tw4l committed:
Remove unused import

tw4l committed:
Default ProfileCreate crawlerid to latest

tw4l committed:
Fix linting

tw4l committed:
Fix import paths

tw4l committed:
Use crawlerid: beta for crawler_crawl_id fixture

tw4l committed:
Try to appease linting gods again

tw4l committed:
Add migration to update workflows, profiles, and configmaps

tw4l committed:
Lint fixes for the linter gods

tw4l committed:
WIP: Add crawlerid to config-details component

Ideally we'd want to be fetching and displaying the crawler version
name here, as that's what we're showing in the new/edit workflow
screens.

tw4l committed:
Add crawler version name/label to config and crawl details

tw4l committed:
Tweak crawler name/label display

tw4l committed:
Fix linting

tw4l committed:
Fix typing for get_crawler_version_by_id

tw4l committed:
WIP: Rebase on latest - select not showing

tw4l committed:
Fixups

- Switch SelectCrawler to subclass from LitElement
- Add missing crawler_image arg to readd_configmap

tw4l committed:
Export select-crawler component

tw4l committed:
Move crawler version above crawl id in crawl details

tw4l committed:
Add backend test for crawler-versions API endpoint

tw4l committed:
Fix test

ikreymer commented:
Some initial thoughts:
- Perhaps should show the full image version as well, either in the dropdown on the right or when selected underneath?
might make it easier to debug in the future.
- Does it make sense to have separate endpoint for this vs just adding it to /api/settings?

ikreymer commented:
Should the `crawlerLabel` be stored in the data model on crawls / crawlconfigs, as well as the crawlerId? It seems like that's just used for displaying something to the user, if so, maybe we want to just look that up on the fly?

Shrinks99 commented:
Agreed that we should show version numbers in the dropdown, this will also be especially valuable info in the crawl settings page!

That said, adding version numbers might also complicate how we name this.  Do we plan to allow users to select a specific version of the crawler or can they only select between beta and stable)?  We don't want to give them the wrong impression that the version number is fixed, maybe this can be handled with help text, changing the label to something like "Release Train", or perhaps this gets relegated to being mentioned in the docs?

ikreymer committed:
additional refactor:
- remove crawlerLabel
- load crawler versions once on init
- store crawlerId in crawlconfig and crawl
- lookup crawler image from crawlerId in operator, to allow updating image dynamically while crawl is running
- rename to crawlerId for consistency on frontend and backend
- frontend: update to use new LiteElement
- frontend: show currently selected version as help tooltip
- fix updating CRAWLER_ID in configmap

ikreymer committed:
additional cleanup:
- switch to dict form for crawler version ids
- fix tests, make configs consistent
- add current crawlerImage to crawl status

ikreymer committed:
rename crawlerId -> crawlerChannel on backend and frontend for clarity!
update types

ikreymer committed:
more renaming CrawlVersion -> CrawlChannel, rename api endpoints
switch json config back to list to preserve order
ui: hide dropdown if only one channel, reset to 'latest' if current channel not found

ikreymer committed:
fix label

ikreymer commented:
Did some refactoring on this, notably:
- Renamed to 'crawler channel' using crawlerChannel on backend and frontend
- Simplified model, removing crawlerLabel, just showing the channel id capitalized as label.
- Crawl stores the channel and is looked up to image in operator (channel->image map is loaded once on startup). This can support dynamically updating the crawler image (either manually or automatically) via a crawl is running if the channel image is updated.
- Added display of current version for each channel:
<img width="902" alt="Screenshot 2024-01-11 at 11 49 03 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/1015759/2e7e010a-1257-4f67-87cc-54e32a264ea3">

- If only one channel, the dropdown is hidden. When saving crawl workflow, invalid channels are automatically defaulted to 'latest'
- Error on startup in `latest` channel does not exist.

ikreymer committed:
error on startup if 'latest' channel is missing

Shrinks99 commented on frontend/src/components/ui/select-crawler.ts:
```diff
@@ -0,0 +1,172 @@
+          ? html`
+              <div
+                slot="help-text"
+                style="font-size: smaller"
```
`help-text` Already has a consistent font size `var(--sl-font-size-x-small)` and shouldn't be adjusted.

ikreymer replied:
yeah, but this isn't help text, but more like debugging / version info, so I think makes sense to make it slightly smaller here.

Shrinks99 replied:
It's supplementary information about the chosen selection just like all the other components using left aligned information in the `help-text` slot?  I don't think it's different enough to warrant a different style.  Seems inconsistently applied here and nowhere else.

Same for below + unlike other right aligned text, it is not explaining limits that the user must understand when editing the form component.

ikreymer replied:
Updated in latest

Shrinks99 commented on frontend/src/components/ui/select-crawler.ts:
```diff
@@ -0,0 +1,172 @@
+              <div
+                slot="help-text"
+                style="font-size: smaller"
+                class="text-right"
```
Right aligned help text in forms is currently used for making the user aware of character limits.  We've used left aligned text to deliver supplemental information about the selected feature.  Would remove this.

ikreymer replied:
imo it looks slightly better, and makes sense, as its supplemental info (like the character limit) and not help text about the field.

Shrinks99 commented on frontend/src/components/ui/select-crawler.ts:
```diff
@@ -0,0 +1,172 @@
+                style="font-size: smaller"
+                class="text-right"
+              >
+                Version:
```
Missing localization wrapper,

```suggestion
              ${msg("Version:")}
```

Shrinks99 commented:
Would rename "latest" to "stable" or similar?  Technically dev would be the "latest" version to receive updates.

ikreymer committed:
keep track of 'image' in crawl (backend and frontend) to be able to display to users
don't auto-update image when channel image changes, just use the one in status
display image in config and crawl detail
make style consistent as per review request
fix typo in configmap, CRAWLER_ID -> CRAWLER_CHANNEL

ikreymer committed:
rename default crawler channel to 'default' instead of 'latest'

Shrinks99 committed:
Adds release channel docs info

Shrinks99 reviewed (approved):
Looks good!  Tested prior to frontend name change, seems to work as expected.  Added docs info for the setting.

Approved! :)

tw4l commented:
Looks great and love the changes! I manually tested:

- Various combinations of configuring and misconfiguring `crawler_channels`
- That dropdowns are present for workflows and profiles only if more than one crawler channel is available
- That the specific crawler version is written into crawls

I had no trouble building the frontend, I think the Frontend Build Check failure might not be related to this PR but a reflection of something in recent main. @emma-sg, would you mind taking a look? (confirmed, can see same failure off of latest main here: https://github.com/webrecorder/browsertrix-cloud/actions/runs/7533278131/job/20505496470?pr=1469; perhaps we need to exclude the frontend `tests` directory from the webpack build?)
